code/modules/power/apc.dm

```diff
@@ -1271,114 +1271,111 @@
 	var/last_eq = equipment
 	var/last_en = environ
 	var/last_ch = charging
-
 	var/excess = surplus()
-
-	if(!src.avail())
+	if(!avail())
 		main_status = 0
 	else if(excess < 0)
 		main_status = 1
 	else
 		main_status = 2
 
-	if(cell && !shorted)
-		// draw power from cell as before to power the area
-		var/cellused = min(cell.charge, GLOB.CELLRATE * lastused_total)	// clamp deduction to a max, amount left in cell
-		cell.use(cellused)
+	var/cur_excess = excess
+	var/cur_used = lastused_total
 
-		if(excess > lastused_total)		// if power excess recharge the cell
-										// by the same amount just used
-			cell.give(cellused)
-			add_load(cellused/GLOB.CELLRATE)		// add the load used to recharge the cell
+	// first: if we have enough power, power the essentials DIRECTLY
 
+	var/environ_satisfied = FALSE
+	var/equipment_satisfied = FALSE
+	var/lighting_satisfied = FALSE
 
-		else		// no excess, and not enough per-apc
-			if((cell.charge/GLOB.CELLRATE + excess) >= lastused_total)		// can we draw enough from cell+grid to cover last usage?
-				cell.charge = min(cell.maxcharge, cell.charge + GLOB.CELLRATE * excess)	//recharge with what we can
-				add_load(excess)		// so draw what we can from the grid
-				charging = APC_NOT_CHARGING
+	if(cur_excess >= lastused_environ)
+		autoset(environ, 1)
+		add_load(lastused_environ)
+		cur_excess -= lastused_environ
+		cur_used -= lastused_environ
+		environ_satisfied = TRUE
 
-			else	// not enough power available to run the last tick!
-				charging = APC_NOT_CHARGING
-				chargecount = 0
+	if(cur_excess >= lastused_equip)
+		autoset(equipment, 1)
+		add_load(lastused_equip)
+		cur_excess -= lastused_equip
+		cur_used -= lastused_equip
+		equipment_satisfied = TRUE
+
+	if(cur_excess >= lastused_light)
+		autoset(lighting, 1)
+		add_load(lastused_light)
+		cur_excess -= lastused_light
+		cur_used -= lastused_light
+		lighting_satisfied = TRUE
+
+
+	// next: take from or charge to the cell, depending on how much is left
+	if(cell && !shorted)
+		if(cur_excess > 0)
+			var/charging_cell = min(cur_excess, cell.maxcharge * GLOB.CHARGELEVEL)
+			cell.give(charging_cell)
+			add_load(charging_cell)
+			lastused_total += charging_cell
+			longtermpower = min(10,longtermpower + 1)
+			if(chargemode && !charging)
+				chargecount++
+				if(chargecount == 10)
```

Comment:
```suggestion
				if(++chargecount == 10)
```

Reply:
Okay for once I'm actually going to say "no, that's disgusting". There's a chance that this might actually make it run faster because byond is not in any way sane, but putting a comparison and a ++ in the same line is, to be frank, fucking disastrous for readability and I'd summarily reject it in any other language. Hell, I'd remove ++ from this language if I could, I don't know why I used it here except that people would complain if I didn't.

Reply:
You are right. I didn't find anything worth to note about so I just dropped that stupid suggestion.